system-addon/content-src/lib/init-store.js

```diff
@@ -76,6 +76,29 @@ const rehydrationMiddleware = store => next => action => {
   return next(action);
 };
 
+// This middleware queues up all the SAVE_SESSION_PERF_DATA actions until it
+// receives the MERGE_STORE_ACTION, which is guaranteed to come from chrome.
+const queueEarlyPerfActionMiddleware = store => next => action => {
+  if (store._ready) {
+    return next(action);
+  }
+
+  if (action.type === MERGE_STORE_ACTION) {
+    next(action);
+    store._ready = true;
+    if (store._queue && store._queue.length) {
+      store._queue.forEach(next);
+    }
+    return null;
+  } else if (action.type === at.SAVE_SESSION_PERF_DATA) {
+    store._queue = store._queue || [];
+    store._queue.push(action);
+    return null;
+  }
+
+  return next(action);
```

Comment:
I believe part of the refactoring was to remove the `return next()` to avoid the confusion of needing to return some value.